x-pack/plugins/task_manager/server/task_claimers/strategy_mget.ts

```diff
@@ -0,0 +1,314 @@
+    ) {
+      currentTasks.add(searchDoc);
+      continue;
+    } else {
+      staleTasks.add(searchDoc);
+      continue;
+    }
+  }
+
+  for (const doc of missingTasks) {
+    logger.info(`Task ${doc.id} is missing from the task store or something`, logMeta);
+  }
+
+  for (const doc of staleTasks) {
+    logger.info(`Task ${doc.id} is stale`, logMeta);
+  }
+
+  const candidateTasks = applyLimitedConcurrency(currentTasks, batches);
+  const taskUpdates: ConcreteTaskInstance[] = Array.from(candidateTasks)
+    .map((task) => {
+      if (task.retryAt != null && new Date(task.retryAt).getTime() < Date.now()) {
+        task.scheduledAt = task.retryAt;
+      } else {
+        task.scheduledAt = task.runAt;
+      }
+      task.retryAt = claimOwnershipUntil;
+      task.ownerId = taskStore.taskManagerId;
+      task.status = TaskStatus.Claiming;
+
+      return task;
+    })
+    .slice(0, initialCapacity);
+
+  const finalResults: ConcreteTaskInstance[] = [];
+  let conflicts = staleTasks.size;
+  try {
+    const updateResults = await taskStore.bulkUpdate(taskUpdates, { validate: true });
+    for (const updateResult of updateResults) {
+      if (isOk(updateResult)) {
+        finalResults.push(updateResult.value);
+      } else {
+        conflicts++;
+        const { id, type, error } = updateResult.error;
+        logger.warn(`Error updating task ${id}:${type} during claim: ${error.message}`, logMeta);
+      }
+    }
+  } catch (err) {
+    logger.warn(`Error updating tasks during claim: ${err}`, logMeta);
+  }
+
+  return {
+    stats: {
+      tasksUpdated: finalResults.length,
+      tasksConflicted: conflicts,
+      tasksClaimed: finalResults.length,
+    },
+    docs: finalResults,
+  };
+}
+
+interface SearchAvailableTasksResponse {
+  docs: ConcreteTaskInstance[];
+  versionMap: Map<string, ConcreteTaskInstanceVersion>;
+}
+
+async function searchAvailableTasks({
+  definitions,
+  excludedTaskTypes,
+  taskStore,
+  claimOwnershipUntil,
+  size,
+  taskTypes,
+  unusedTypes,
+  taskMaxAttempts,
+}: OwnershipClaimingOpts): Promise<SearchAvailableTasksResponse> {
+  // TODO: handle excludedTaskTypes
+  // const { taskTypesToSkip = [], taskTypesToClaim = [] } = groupBy(
+  //   definitions.getAllTypes(),
+  //   (type) =>
+  //     taskTypes.has(type) && !isTaskTypeExcluded(excludedTaskTypes, type)
+  //       ? 'taskTypesToClaim'
+  //       : 'taskTypesToSkip'
+  // );
+  const queryForScheduledTasks = mustBeAllOf(
+    // Task must be enabled
+    EnabledTask,
+    // Either a task with idle status and runAt <= now or
+    // status running or claiming with a retryAt <= now.
+    shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt)
+  );
+
+  const sort: NonNullable<SearchOpts['sort']> = getClaimSort(definitions);
+  const query = matchesClauses(queryForScheduledTasks, filterDownBy(InactiveTasks));
+
+  const apmTrans = apm.startTransaction(
```

Comment:
We'll need to move this within `claimAvailableTasks` so the transaction includes mget and bulkUpdate. Here's how I fixed it in my PoC: https://github.com/elastic/kibana/pull/182394/commits/2eb0b9a212f1686a5d6d2a0b283fb0319d14e56a?diff=unified&w=1

Reply:
Nice!